### ISSUE TITLE
data.batch appears to be None

### ISSUE THREAD

**Issue by legohyl** [rusty1s/pytorch_geometric#1272]
## ❓ Questions & Help

I'm trying to use the `global_max_pooling` function and it requires the `batch` vector. The thing is, oddly, my `data.batch` seems to be of `None` type. Is this due to an implementation error when I implemented my own `Data` object? I implemented the `__inc__` function when declaring my own `Data` object.

**Comment by rusty1s** [comment]
I see. PyG does only know how to create the `batch` vector if it can figure out the number of nodes in your graph. You can force the `DataLoader` to create batch vectors though:
```python
loader = DataLoader(dataset, batch_size=32, follow_batch=['name_of_attribute'])
```
where `name_of_attribute` corresponds to the name of the attribute in `data` you want to separate, *e.g.*, `x`.

**Comment by legohyl** [comment]
Oh I see. Does this mean I have to implement a `num_nodes` kind of value in the `Data` object?

**Comment by rusty1s** [comment]
That's hard to say without knowing your data. Is it some kind of bipartite graph? Is it a pair of graphs?

So you have two options:
(1) Set `data.num_nodes = ...` on each `data` object.
(2) Create batch vectors based on some attributes in your data.

The (1) option might be the easiest one to try.

**Comment by legohyl** [comment]
Yup, it's a bipartite graph. I am tracking it as `node_set_a` and `node_set_b`. To implement the `__inc__` method,  I used `node_set_a.size(0) + node_set_b.size(0)`.

**Comment by rusty1s** [comment]
```python
loader = DataLoader(dataset, batch_size=32, follow_batch=['node_set_a', 'node_set_b'])
```
should work just fine, giving you two batch vectors for each batched node set.

**Comment by legohyl** [comment]
Hmm from what I understand, `__inc__` tells the `DataLoader` how much to increment each of the indexes by because we stack all the nodes together in one large disconnected graph with a large adjacency matrix right? 

Hence, I'd like to ask what does the `follow_batch` do for the DataLoader? If I were trying to `graph_max_pooling` on one graph in my batch, I should feed it both batch vectors so that it has. all the nodes for that graph?

**Comment by rusty1s** [comment]
If you have a bipartite graph, I guess you want to learn node embeddings for a node set given messages received from the other node set? Like learn node embeddings for `node_set_b` given messages from `node_set_a`. You can then max pool node embeddings from `node_set_b` via
```python
global_max_pool(node_set_b, batch=data.batch_node_set_b)
```

**Comment by legohyl** [comment]
Haha, not exactly so. I actually want to a `global_max_pool` on the entire graph. So `node_set_a` and `node_set_b` are bipartite, what I'm trying to do is the passage messages between them, same like belief propagation algorithm. Now, I wish to do a graph pool to get an overall representation of this graph, using max-pooling.

**Comment by rusty1s** [comment]
I see. In that case, you don't actually need to separate node_set_a and node_set_b features, and you don't need to modify `__inc__`. You can just implement your graph with `node_size = node_set_a.size(0) + node_set_b.size(0)`, where edges only exist between nodes from different sets. The `batch` vector should be generated automatically.

**Comment by legohyl** [comment]
Ah okay, alright I'll give that a go. Thanks a lot!

**Comment by legohyl** [comment]
Hey Matthias, sorry to re-open this, I managed to get the `num_nodes` working and that does indeed create the `Batch` vector. However, I'd like repeat the output of the `global_max_pool` and concat all my node features. Since the result of `global_max_pool` is of shape `b x f`, where `b` is the batch and `f` is the feature set size, how do I repeat the vector enough times to concat it? Meaning that my data is a batch now is something like batch_size = 2, batch1_1 = 10nodes, batch1_2 = 20nodes. 

I'd like to concat my `global_max_pool` vector such that it should be 10 repeats of [1, f] and 20 repeats of [2, f], followed by concat-ing them to the node feature values.

**Comment by rusty1s** [comment]
```python
x_global = global_max_pool(x, batch)
x = torch.cat([x, x_global[batch]], dim=-1)
```

**Comment by legohyl** [comment]
Oh wow, thanks, that's really quick. Could you please explain the logic behind that?
`x_global[batch]` creates sufficient copies according to the correct batch label, how does the `torch.cat` along `dim=-1` infer the correct axis?

**Comment by rusty1s** [comment]
It concats the global feature matrix and the individual node features (last dimension), so the axis should be correct.

**Comment by legohyl** [comment]
Cheers and thanks for the explanation, really helped a lot!
